Implement `debug_getRawReceipts`

tcoratger opened the pull request:
<!--- Please provide a general summary of your changes in the title above -->

Time spent on this PR: 2h

Resolves: #524 

# Pull Request type

<!-- Please try to limit your pull request to one type; submit multiple pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?

<!-- Please describe the behavior or changes that are being added by this PR. -->

- Updated Reth version to `v0.2.0-beta.3` to access the `envelope_encoded` method on `ReceiptWithBloom` type.
- Implemented the `raw_receipts` method to retrieve receipts as a vector of bytes for a given block ID, which can be a hash or a number. Utilized existing functions and transformed the `TransactionReceipt` type outputted from the `block_receipts` method into a `Receipt` type, as all necessary information is contained therein. Considered alternative methods for capturing single block transaction indices, but decided against them due to potential complexity and impact on the indexer.
- Added unit tests to ensure proper functionality with both block numbers and block hashes.
- Also added a `block_by_id` method, initially intended for use but ultimately unnecessary. Left in place for potential future use, providing flexibility to switch between `block_by_hash` and `block_by_number` depending on the provided block ID.

# Does this introduce a breaking change?

- [ ] Yes
- [X] No

greged93 commented on src/eth_provider/provider.rs:
```diff
@@ -5,11 +5,13 @@ use cainome::cairo_serde::CairoArrayLegacy;
 use eyre::Result;
 use itertools::Itertools;
 use mongodb::bson::doc;
+use reth_primitives::serde_helper::JsonStorageKey;
+use reth_primitives::serde_helper::U64HexOrNumber;
```
Can we just group these in one import?

tcoratger replied:
Done. I think in the future we can use this

https://github.com/paradigmxyz/reth/blob/6801869c0d3e5ebf1b187d64694ec7ca13d809ea/rustfmt.toml#L2

in order to automatically classify the imports on save.

greged93 commented on src/eth_rpc/api/eth_api.rs:
```diff
@@ -1,11 +1,12 @@
 use jsonrpsee::core::RpcResult as Result;
 use jsonrpsee::proc_macros::rpc;
+use reth_primitives::serde_helper::JsonStorageKey;
+use reth_primitives::serde_helper::U64HexOrNumber;
```
also group here

tcoratger replied:
Done

greged93 commented on src/eth_rpc/servers/debug_rpc.rs:
```diff
@@ -60,7 +59,27 @@ impl<P: EthereumProvider + Send + Sync + 'static> DebugApiServer for DebugRpc<P>
+            .into_iter()
+            .map(|receipt| {
+                Receipt {
+                    tx_type: receipt.transaction_type.to::<u8>().try_into().unwrap(),
```
Can we avoid the unwrap?

greged93 commented on src/eth_rpc/servers/debug_rpc.rs:
```diff
@@ -60,7 +59,27 @@ impl<P: EthereumProvider + Send + Sync + 'static> DebugApiServer for DebugRpc<P>
+            .map(|receipt| {
+                Receipt {
+                    tx_type: receipt.transaction_type.to::<u8>().try_into().unwrap(),
+                    success: receipt.status_code.unwrap_or_default().to::<u8>() == 1,
```
the `to` method can panic, better to either use `TryInto::<u8>` or there is also a `saturating_to`

tcoratger replied:
Here as `status_code` is a `U64` I simply converted it with the to method to `u64`, which is safe. In any case, if the conversion does not give 1 it is ultimately false.

greged93 commented on src/eth_rpc/servers/debug_rpc.rs:
```diff
@@ -60,7 +59,27 @@ impl<P: EthereumProvider + Send + Sync + 'static> DebugApiServer for DebugRpc<P>
+                Receipt {
+                    tx_type: receipt.transaction_type.to::<u8>().try_into().unwrap(),
+                    success: receipt.status_code.unwrap_or_default().to::<u8>() == 1,
+                    cumulative_gas_used: receipt.cumulative_gas_used.to::<u64>(),
```
Same, I would in general avoid using `to` unless when converting from `U_n` to `u_n`.

tcoratger replied:
I've used the same approach as before here, with the error handled into a conversion error item.

greged93 replied:
Maybe indeed, misread the code

greged93 commented on src/eth_rpc/servers/debug_rpc.rs:
```diff
@@ -60,7 +59,27 @@ impl<P: EthereumProvider + Send + Sync + 'static> DebugApiServer for DebugRpc<P>
+                        .map(|log| Log { address: log.address, topics: log.topics, data: log.data })
+                        .collect(),
+                }
+                .with_bloom()
```
`with_bloom` is quite a heavy computation, we actually store the receipts in our db with the blooms, so you could just get the bloom from receipt.bloom and avoid this computation.

tcoratger replied:
Yes right, just modified that to integrate the bloom directly without going through this slow method.

greged93 commented on src/eth_rpc/servers/eth_rpc.rs:
```diff
@@ -1,10 +1,12 @@
 #![allow(clippy::blocks_in_conditions)]
 
 use jsonrpsee::core::{async_trait, RpcResult as Result};
+use reth_primitives::serde_helper::JsonStorageKey;
+use reth_primitives::serde_helper::U64HexOrNumber;
```
can we group these

tcoratger replied:
Done

greged93 commented on tests/debug_api.rs:
```diff
@@ -186,3 +186,70 @@ async fn test_raw_transaction(#[future] katana: Katana, _setup: ()) {
+
+    let rlp_bytes_by_block_number: Vec<Bytes> =
+        serde_json::from_value(raw_by_block_number["result"].clone()).expect("Failed to deserialize result");
+
```
can you just add an assertion based a little on what was done in `test_raw_transaction`, to assert that the rlp encoded receipt corresponds to the receipt we get with `getTransactionReceipt`?

tcoratger replied:
Yes so I did a little differently here, tell me what you think. As during `transaction_receipt`, we obtain a different type `TransactionReceipt` which is an rpc type unlike `ReceiptWithBloom` which is a primitive reth type, I used the following method:
- I opened an eth provider.
- I caught all receipts for a block with `block_receipts` to be iso receipts.
- For each receipt, I extracted the hash of the transaction and I tried to look for the receipt of this transaction using the alternative method `transaction_receipt`.
- Then I transform the result which is `TransactionReceipt` into a `ReceiptWithBloom`.
- Finally, I compare the result of my rpc `debug_getRawReceipts` with each receipt obtained for the block.

This way I make sure that each receipt in the block is correct, tell me what you think, I think it's a good way to check the consistency of the two methods.

greged93 replied:
that works for me yes

greged93 commented:
Hey @tcoratger thanks for the PR! Couple of comments (mostly nits), and concerning `block_by_id`, I think we can remove it for now, we try to follow [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)

greged93 commented:
just realized we might need to update `start_kakarot_rpc_server` helper function to spawn the rpc server on different `ports` or the integration tests won't work I think

tcoratger commented:
> Hey @tcoratger thanks for the PR! Couple of comments (mostly nits), and concerning `block_by_id`, I think we can remove it for now, we try to follow [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it). I tried some tests locally and they it seems they weren't compiling yet (some imports issues). Just in case: for running our tests locally, you need a docker instance running and you can run the full suite with `make test`. If you want to run only one test, you need to run `make katana-genesis` once first (generates a genesis for all the tests) and then `make test-target TARGET=your_test`

- I've just removed `block_by_id` method.
- In fact for the tests, I only ran mine locally and I had not checked the existing ones. But because of the update of the reth version, there are a few types that changed and as a result it caused the tests to fail, I corrected that and run locally normally it works, don't hesitate to tell me if necessary.

Eikix commented on src/eth_rpc/config.rs:
```diff
@@ -14,4 +14,14 @@ impl RPCConfig {
         Ok(Self::new(socket_addr))
     }
+
+    pub fn from_port(port: u16) -> Result<Self> {
```
thanks a lot, we've been needing this for a long time, was going to impl. it for debug_rawBlock 

if we wanted to go one step further we could decouple testing environment from actual .env file, it feels wrong that our tests rely on non-committed .env files, 
But since it probably belongs in a smol non-related PR, could you create an issue?

I suggest adding to the Config struct a new method `::new_test_config()` method, or something similar, under `testing` feature. Let me know if you have a better idea 👑

tcoratger replied:
Good idea, just opened a PR here: https://github.com/kkrt-labs/kakarot-rpc/issues/879

greged93 reviewed (dismissed):
lgtm!

Eikix commented:
Great work thanks 🥕👑

Eikix commented:
clippy and good to go

tcoratger commented:
> clippy and good to go

@Eikix Yes just fixed

greged93 reviewed (approved):
lgtm!
